lib/repl.js

```diff
@@ -949,7 +951,32 @@ function REPLServer(prompt,
           self.displayPrompt();
           return;
         }
+      }
+
+      // In the next two if blocks, we do not use os.EOL instead of '\n'
+      // because on Windows it is '\r\n'
+      if (cmd && StringPrototypeIndexOf(cmd, '\n') !== -1) { // If you are editing a multiline command
+        self.history[0] = StringPrototypeReplaceAll(cmd, '\n', '\r');
+      } else if (self[kBufferedCommandSymbol]) { // If a new multiline command was entered
+        // Remove the first N lines from the self.history array
+        // where N is the number of lines in the buffered command
+
+        const lines = StringPrototypeSplit(self[kBufferedCommandSymbol], '\n');
+        self.history = ArrayPrototypeSlice(self.history, lines.length);
+        ArrayPrototypePop(lines);
+        // And replace them with the single command split by '\r'
+        ArrayPrototypePush(lines, cmd);
+        const newHistoryLine = ArrayPrototypeJoin(lines, '\r');
+        if (self.history[0] !== newHistoryLine) {
```

Comment:
I believe this prevents duplicate entries?

Since we already have such functionality somewhere, should that not be adjusted to just also be able to handle multiline input?

Reply:
That's what I was trying in the beginning, but the problem is that each line ending with a `\n` is added to the history independently, calling `[kAddHistory]()`. The `reconstruction` of the history from a "bunch of single lines" to an actual multiline is done in this highlighted function

Reply:
kAddHistory should likely be part of the repl and not of readline (or we should expose the API regularly on readline. I am just not sure if that is really something that belongs to readline).

There is definitely need for clearer separation of the two modules. While I understand that it's out of scope for the PR.